### PR TITLE
feat(Bonus Pagamenti Digitali): [#175949561] Display cashback future periods ("Inactive") in wallet

### PR DESCRIPTION
**Pull request by fabriziofff**
## Short description
This pr allows to see in the wallet a cashback card for the first "Inactive" period, if the user is enrolled to bpd and all the periods are inactive.

1 closed period - 1 active period - 2 inactive periods => 2 inactive periods are hidden

2 inactive periods => The first inactive period is shown.

### CODE REVIEW

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+      expect(pot.isSome(visiblePeriods)).toBeTruthy();
+      if (pot.isSome(visiblePeriods)) {
+        expect(visiblePeriods.value.length).toBe(0);
+      }
+    }
+  );
+});
+
+describe("test bpdPeriodsAmountWalletVisibleSelector when bpd is disabled", () => {
+  it("one inactive period should return one inactive period", () => {
```
Comment (CrisTofani) [inline ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:204]:
```suggestion
  it("one inactive period should return no periods", () => {
```

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+      pot.some([inactivePeriodB]),
+      remoteReady(false)
+    );
+
+    expect(pot.isSome(visiblePeriods)).toBeTruthy();
+    if (pot.isSome(visiblePeriods)) {
+      expect(visiblePeriods.value.length).toBe(0);
+    }
+  });
+  it("with multiple inactive period should return the most recent inactive period", () => {
```
Comment (CrisTofani) [inline ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:215]:
```suggestion
  it("with multiple inactive period should return no periods", () => {
```

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+
+    expect(pot.isSome(visiblePeriods)).toBeTruthy();
+    if (pot.isSome(visiblePeriods)) {
+      expect(visiblePeriods.value.length).toBe(0);
+    }
+  });
+
+  it(
+    "with multiple inactive period and an active period " +
+      "should return the active period",
```
Comment (CrisTofani) [inline ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:229]:
```suggestion
      "should return no periods",
```

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      expect(pot.isSome(visiblePeriods)).toBeTruthy();
+      if (pot.isSome(visiblePeriods)) {
+        expect(visiblePeriods.value.length).toBe(0);
+      }
+    }
+  );
+
+  it(
+    "with multiple inactive period, an active period and a closed period (with amount zero) " +
+      "should return the active period",
```
Comment (CrisTofani) [inline ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:250]:
```suggestion
      "should return no periods",
```

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+      expect(pot.isSome(visiblePeriods)).toBeTruthy();
+      if (pot.isSome(visiblePeriods)) {
+        expect(visiblePeriods.value.length).toBe(0);
+      }
+    }
+  );
+
+  it(
+    "with multiple inactive period, an active period and a closed period (with amount > zero) " +
+      "should return the active period and the closed period",
```
Comment (CrisTofani) [inline ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:272]:
```suggestion
      "should return the closed period",
```

**Review by CrisTofani — CHANGES_REQUESTED** [review]

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts (hunk 1 above)
Comment (fabriziofff) [reply ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:204]: When there is one inactive (future) period, that period should be returned, following the new logic.

File: ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts (hunk 2 above)
Comment (fabriziofff) [reply ts/features/bonus/bpd/store/reducers/__test__/combiner.test.ts:215]: When there are multiple inactive (future) period, that most recent period should be returned, following the new logic.

**Review by CrisTofani — APPROVED** [review]

File: ts/features/bonus/bpd/store/reducers/details/combiner.ts [hunk 6]
```diff
@@ -49,31 +49,54 @@ export const bpdAllPeriodsWithAmountSelector = createSelector(
...
 ) =>
   isReady(bpdEnabled) &&
   ((periodAmount.period.status === "Active" && bpdEnabled.value) ||
     (periodAmount.period.status === "Closed" &&
-      periodAmount.amount.transactionNumber > 0));
+      periodAmount.amount.transactionNumber > 0) ||
+    // All the periods are inactive
+    (periodList.every(p => p.period.status === "Inactive") &&
+      // This is the first inactive period
+      periodList.indexOf(periodAmount) === 0 &&
```
Comment (Undermaken) [inline ts/features/bonus/bpd/store/reducers/details/combiner.ts:71]: This is a shallow equality: it checks about object reference.

```typescript

const inactivePeriodA: BpdPeriodAmount = {
  amount: zeroAmount,
  period: {
    ...inactivePeriod,
    startDate: new Date("2025-01-01"),
    awardPeriodId: 55 as AwardPeriodId
  }
};

const test = isPeriodAmountWalletVisible(
      [inactivePeriodA],
      inactivePeriodA,
      remoteReady(true)
    );

    const test2 = isPeriodAmountWalletVisible(
      [inactivePeriodA],
      { ...inactivePeriodA },
      remoteReady(true)
    );
```
`test -> true`
`test2 -> false`

A solution could be

```typescript
import _ from "lodash";
periodList.findIndex(o => _.isEqual(o, periodAmount)) === 0 
```

File: ts/features/bonus/bpd/store/reducers/details/combiner.ts [hunk 7]
```diff
@@ -49,31 +49,54 @@ export const bpdAllPeriodsWithAmountSelector = createSelector(
...
-      periodsAmountList.filter(periodAmount =>
-        isPeriodAmountWalletVisible(periodAmount, bpdEnabled)
-      )
-    )
+    pot.map(potPeriodsAmount, periodsAmountList => {
+      const periodsOrderedByDate = periodsAmountList
+        // create a sorted copy of the array
+        .concat()
+        .sort((pa1, pa2) =>
+          pa1.period.startDate < pa2.period.startDate
```
Comment (Undermaken) [inline ts/features/bonus/bpd/store/reducers/details/combiner.ts:86]:
```suggestion
import _ from "lodash";
_.clamp(pa1.period.startDate - pa2.period.startDate, -1, 1);
```
could it be a refactoring?

**Review by Undermaken — CHANGES_REQUESTED** [review]

**Pull request merged**